## Step 1: Import Libraries

### 📚 Your Task:
Import the necessary libraries for data analysis and machine learning.

### 💡 Hint:
You'll need:
- Data manipulation: `pandas`, `numpy`
- Visualization: `matplotlib`, `seaborn`
- ML tools: `sklearn` (various modules)

### ✍️ Write your code below:

In [2]:
# TODO: Import all necessary libraries
# Your code here:
import pandas as pd
import numpy as np
import polars as pl
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder, LabelEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import warnings
warnings.filterwarnings('ignore')

# Set plotting style
plt.style.use('default')
sns.set_palette("husl")

print("📚 All libraries imported successfully!")
print("🔧 Environment configured for machine learning!")

📚 All libraries imported successfully!
🔧 Environment configured for machine learning!


## Step 2: Load the Dataset

### 📚 Your Task:
Load the Adult Income dataset from `adult.csv`

### 💡 Hints:
- Use `pd.read_csv()` to load the data
- Check the first few rows with `.head()`
- Examine the shape with `.shape`
- Look at data types with `.dtypes`
- Check for missing values with `.isnull().sum()`

### ✍️ Write your code below:

In [3]:
# TODO: Load the dataset
# Your code here:

df = pd.read_csv("adult.csv")
print("Dataset skibidi")



Dataset skibidi


In [4]:
# TODO: Display basic information about the dataset
# Your code here:

print(f"Shape of Datashape {df.shape}")

print("\nData Types:")
print(df.dtypes)

print("\nMissing Values:")
print(df.isnull().sum())

print("\nFirst 5 rows:")
display(df.head())

Shape of Datashape (48842, 15)

Data Types:
age                 int64
workclass          object
fnlwgt              int64
education          object
educational-num     int64
marital-status     object
occupation         object
relationship       object
race               object
gender             object
capital-gain        int64
capital-loss        int64
hours-per-week      int64
native-country     object
income             object
dtype: object

Missing Values:
age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64

First 5 rows:


,age,workclass,fnlwgt,education,educational-num,marital-status,occupation,relationship,race,gender,capital-gain,capital-loss,hours-per-week,native-country,income
0,25,Private,226802,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K
1,38,Private,89814,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States,<=50K
2,28,Local-gov,336951,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States,>50K
3,44,Private,160323,Some-college,10,Married-civ-spouse,Machine-op-inspct,Husband,Black,Male,7688,0,40,United-States,>50K
4,18,?,103497,Some-college,10,Never-married,?,Own-child,White,Female,0,0,30,United-States,<=50K




### 💡 Visualization Ideas:
- Histograms for numerical features
- Bar charts for categorical features
- Box plots to compare groups
- Scatter plots for relationships
- Correlation heatmaps

### ✍️ Write your code below:

In [ ]:
# TODO: Analyze the target variable distribution
# Your code here:



In [ ]:
# TODO: Create visualizations to explore numerical features
# Your code here:


In [27]:
# TODO: Explore categorical features
# Your code here:
df_pl = pl.from_pandas(df)
("3capital-gain and capital loss by marital-status")
q3 = df_pl.group_by("marital-status").agg([
    pl.col("capital-gain").sum().alias("total_cap_gain"),
    pl.col("capital-loss").sum().alias("total_cap_loss")
])
print(q3)
("5 which occupation has the highest average age?")
q5 = df_pl.group_by("occupation").agg(pl.col("age").mean().alias("avg_age")).sort("avg_age",descending=True).head(1)
print(q5)
("7 Do people with higher capital-gain tend to work more hours?") 
q7 = df_pl.select(pl.corr("capital-gain", "hours-per-week").alias("correlation"))
print(q7)

shape: (7, 3)
┌───────────────────────┬────────────────┬────────────────┐
│ marital-status        ┆ total_cap_gain ┆ total_cap_loss │
│ ---                   ┆ ---            ┆ ---            │
│ str                   ┆ i64            ┆ i64            │
╞═══════════════════════╪════════════════╪════════════════╡
│ Never-married         ┆ 6195095        ┆ 872350         │
│ Married-civ-spouse    ┆ 38932760       ┆ 2699344        │
│ Widowed               ┆ 916332         ┆ 123900         │
│ Married-spouse-absent ┆ 395015         ┆ 39680          │
│ Married-AF-spouse     ┆ 109950         ┆ 3136           │
│ Separated             ┆ 890219         ┆ 86627          │
│ Divorced              ┆ 5264450        ┆ 448751         │
└───────────────────────┴────────────────┴────────────────┘
shape: (1, 2)
┌─────────────────┬───────────┐
│ occupation      ┆ avg_age   │
│ ---             ┆ ---       │
│ str             ┆ f64       │
╞═════════════════╪═══════════╡
│ Priv-house-serv ┆ 43.400826 │


In [ ]:
# TODO: Investigate relationships between features and target
# Your code here:


### 📝 Your EDA Findings:

**Write your observations here:**
- Key insight 1: The highest is 38932760  married-civ-spouse, the highest capital loss 2699344 married-civ-spouse
- Key insight 2: the average Priv-house-serv age is 43
- Key insight 3: So the capital gain and workhours does not correlate.

- Data quality issues found: 
- Potential challenges for modeling: 

## Step 4: Data Preprocessing

### 📚 Your Tasks:

#### 4.1 Handle Missing Values
- Identify columns with missing values
- Decide on strategy: drop, fill with mean/median/mode, or impute

#### 4.2 Feature Selection
- Decide which features to keep
- Remove irrelevant or redundant columns

#### 4.3 Encode Categorical Variables
- Identify categorical columns
- Choose encoding method (OneHot, Label, etc.)

#### 4.4 Scale Numerical Features
- Identify numerical columns
- Choose scaling method (MinMax, Standard, etc.)

#### 4.5 Prepare Target Variable
- Encode target if needed
- Ensure it's in correct format

### ⚠️ Important Reminders:
- **Split data BEFORE preprocessing** to avoid data leakage
- **Fit transformers on training data only**
- **Use pipelines** for clean, reproducible code

### ✍️ Write your code below:

In [59]:
# TODO: Handle missing values
# Your code here:
print(df.isnull().sum())

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64


In [60]:
# TODO: Separate features (X) and target (y)
# Your code here:
X= df.drop('income', axis=1)
y=df['income']
y = y.map({'<=50K': 0, '>50K' : 1})
   
#Drop irrelevant features


In [63]:
# TODO: Create preprocessing pipeline
# Hint: Use ColumnTransformer to handle different column types
# Your code here:
numberic_features = ['age', 'educational-num', 'capital-gain', 'capital-loss', 'hours-per-week']
categorical_features = ['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'gender', 'native-country']

numeric_transformer = Pipeline(steps=[
    ('scaler', MinMaxScaler())
])

categorical_transformer = Pipeline(steps = [
    ('encoder', OneHotEncoder(drop='first', handle_unknown= 'ignore', sparse_output=False))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numberic_features),
        ('cat', categorical_transformer, categorical_features)
    ]
)

print("Preprocessing pipeline created!")



Preprocessing pipeline created!


### 📝 Your Preprocessing Decisions:

**Document your choices:**
- Missing value strategy: None
- Features selected: All
- Features removed: None
- Categorical encoding method:OneHotEncoder 
- Numerical scaling method: minmaxscaler
- Train/test split ratio: 6,7%-93,3%


## Step 5: Model Building & Training

### 📚 Your Task:
Build and train **at least 2 different** classification models.

### 🤖 Model Options:
1. **Logistic Regression** - Simple, interpretable baseline
2. **Random Forest** - Powerful ensemble method
3. **Decision Tree** - Easy to visualize and understand
4. **K-Nearest Neighbors (KNN)** - Instance-based learning
5. **Support Vector Machine (SVM)** - Effective for classification
6. **Gradient Boosting** - Advanced ensemble technique

### 💡 Tips:
- Start with a simple model (Logistic Regression)
- Then try a more complex model (Random Forest)
- Use pipelines to combine preprocessing and modeling
- Document why you chose each model

### ✍️ Write your code below:

In [ ]:
# TODO: Model 1 - Build and train your first model
# Your code here:
#Logistic Regression
model_1 = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier',LogisticRegression(max_iter=1000, random_state=67))
])
model_1.fit(X=X_train,y=y_train)
print("femboy is logistic Regression trained!")


femboy is logistic Regression trained!


In [ ]:
# TODO: Model 2 - Build and train your second model
# Your code here:
# TODO: Model 1 - Build and train your first model
# Your code here:
#Logistic Regression
model_2 = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier',RandomForestClassifier(n_estimators=6767, random_state=36))
])
model_2.fit(X_train, y_train)
print("femboy Random Forest trained!")


In [ ]:
# TODO: (Optional) Model 3 - Try additional models if you want
# Your code here:
# TODO: Model 2 - Build and train your second model
# Your code here:


### 📝 Your Model Choices:

**Explain your reasoning:**
- Model 1: 
  - Why chosen: 
  - Expected strengths: 
  
- Model 2: 
  - Why chosen: 
  - Expected strengths: 

## Step 6: Model Evaluation

### 📚 Your Task:
Evaluate each model's performance using multiple metrics.

### 📊 Evaluation Metrics to Calculate:
1. **Accuracy** - Overall correctness
2. **Precision** - Of predicted positives, how many were correct?
3. **Recall** - Of actual positives, how many did we find?
4. **F1-Score** - Harmonic mean of precision and recall
5. **Confusion Matrix** - Detailed breakdown of predictions

### 💡 Evaluation Tips:
- Don't rely on accuracy alone (especially with imbalanced data)
- Use `classification_report` for comprehensive metrics
- Visualize confusion matrices
- Consider which metric matters most for this problem

### ✍️ Write your code below:

In [ ]:
# TODO: Evaluate Model 1
# Your code here:
y_pred_1 = model_1.predict(X=X_test)
print("6767 Logistic Regression Perfomance:")
print(classification_report(y_test, y_pred_1))

print("confusion Matrix:")
print(confusion_matrix(y_test, y_pred_1))


6767 Logistic Regression Perfomance:
              precision    recall  f1-score   support

           0       0.88      0.93      0.91      2490
           1       0.74      0.59      0.66       783

    accuracy                           0.85      3273
   macro avg       0.81      0.76      0.78      3273
weighted avg       0.85      0.85      0.85      3273

confusion Matrix:


NameError: name 'y_pred_2' is not defined

In [72]:
# TODO: Evaluate Model 2
# Your code here:

y_pred_2 = model_2.predict(X=X_test)
print("6767 Logistic Regression Perfomance:")
print(classification_report(y_test, y_pred_2))

print("confusion Matrix:")
print(confusion_matrix(y_test, y_pred_2))


6767 Logistic Regression Perfomance:
              precision    recall  f1-score   support

           0       0.87      0.92      0.90      2490
           1       0.69      0.58      0.63       783

    accuracy                           0.84      3273
   macro avg       0.78      0.75      0.76      3273
weighted avg       0.83      0.84      0.83      3273

confusion Matrix:
[[2286  204]
 [ 331  452]]


In [ ]:
# TODO: Create comparison visualizations
# Hint: Compare confusion matrices, plot metric comparisons
# Your code here:


### 📝 Your Evaluation Results:

**Model Performance Summary:**

| Metric | Model 1 | Model 2 | Model 3 (if used) |
|--------|---------|---------|-------------------|
| Accuracy | 
| Precision |0.85|
| Recall | |0.85
| F1-Score | 0.85|

**Analysis:**
- Best performing model: 
- Why this model performed best: 
- Surprising findings: 
- Areas for improvement: 

## Step 6.5: Feature Importance Analysis

### 📚 Your Task:
Find out which features matter most to your model.

### 💡 Why this matters:
Understanding **why** a model makes a prediction is often as important as the prediction itself. 
- If you used a **Decision Tree** or **Random Forest**, you can easily extract `feature_importances_`.
- If you used **Logistic Regression**, you can look at the `coef_` (coefficients).

### ✍️ Write your code below:

In [ ]:
# TODO: Visualize feature importance
# Hint: If using a pipeline with a model named 'classifier', you can access it via:
# model.named_steps['classifier'].feature_importances_
# You'll also need to get feature names from your preprocessor!

# Your code here:


## Step 6.6: Testing on "Real" People

### 🚀 Your Task:
Create a fictional person and ask your model to predict their income category.

### 💡 Instructions:
1. Create a dictionary representing a person (e.g., "Alice, 35, PhD...")
2. Convert it to a DataFrame
3. Pass it to your model's `.predict()` method

### 🧪 Test Case Example:
```python
new_person = pd.DataFrame({
    'age': [35],
    'workclass': ['Private'],
    'education': ['Masters'],
    'marital-status': ['Married-civ-spouse'],
    'occupation': ['Exec-managerial'],
    'relationship': ['Wife'],
    'race': ['White'],
    'gender': ['Female'],
    'capital-gain': [0],
    'capital-loss': [0],
    'hours-per-week': [50],
    'native-country': ['United-States']
})
```

### ✍️ Write your code below:

In [ ]:
# TODO: Create a new person and predict their income
# Your code here:


## Bonus: Save Your Model

### 📚 Your Task:
Save your trained model to a file so it can be used later without retraining.

### 💡 Hint:
Use the `joblib` library.
```python
import joblib
joblib.dump(your_pipeline, 'income_predictor_model.pkl')
```

### ✍️ Write your code below:

In [ ]:
# TODO: Save your best model to a .pkl file
# Your code here:


## Step 7: Model Improvement (Optional Challenge)

### 🚀 Your Task:
Try to improve your best model's performance.

### 💡 Improvement Strategies:

#### 1. Feature Engineering
- Create new features (e.g., age groups, income ratios)
- Combine existing features
- Remove less important features

#### 2. Handle Class Imbalance
- Try class weights: `class_weight='balanced'`
- Use SMOTE for oversampling
- Try undersampling majority class

#### 3. Hyperparameter Tuning
- Use GridSearchCV or RandomizedSearchCV
- Tune model-specific parameters
- Use cross-validation

#### 4. Try Different Preprocessing
- Different scaling methods
- Different encoding strategies
- Feature selection techniques

### ✍️ Write your code below:

In [ ]:
# TODO: Try improvement strategy 1
# Your code here:


In [ ]:
# TODO: Try improvement strategy 2
# Your code here:


In [ ]:
# TODO: Compare improved model with original
# Your code here:


### 📝 Improvement Results:

**What did you try?**
- Strategy 1: 
  - Result: 
  
- Strategy 2: 
  - Result: 

**Final improvement:**
- Original best score: 
- Improved score: 
- Improvement achieved: 

## Step 8: Conclusions & Recommendations

### 📚 Your Task:
Summarize your findings and provide actionable insights.

### 🤔 Questions to Answer:

#### 1. Project Summary
- What was the goal of this project?
- What approach did you take?
- What were the main challenges?

#### 2. Key Findings
- Which features were most important for predicting income?
- What patterns did you discover in the data?
- Were there any surprising results?

#### 3. Model Performance
- Which model performed best and why?
- What were the limitations of your models?
- How confident are you in the predictions?

#### 4. Real-World Application
- How could this model be used in practice?
- What are the ethical considerations?
- What improvements would you make with more time/resources?

#### 5. Lessons Learned
- What did you learn from this project?
- What would you do differently next time?
- What skills did you develop?

### ✍️ Write your conclusions below:

### 📝 Final Project Report:

#### Executive Summary
*Write 2-3 sentences summarizing the entire project*


#### Key Findings
*List your most important discoveries*
1. 
2. 
3. 

#### Model Recommendation
*Which model would you deploy and why?*


#### Limitations & Future Work
*What could be improved?*


#### Personal Reflection
*What did you learn?*


## 🎉 Congratulations!

You've completed your first end-to-end machine learning project!

### ✅ Skills You've Demonstrated:
- ✓ Data loading and exploration
- ✓ Exploratory data analysis
- ✓ Data preprocessing and cleaning
- ✓ Feature engineering
- ✓ Model building and training
- ✓ Model evaluation and comparison
- ✓ Critical thinking and problem-solving
- ✓ Technical communication

### 🚀 Next Steps:
1. **Save your work**: Make sure all outputs are visible
2. **Review your code**: Clean up and add comments
3. **Prepare presentation**: Be ready to explain your approach
4. **Get feedback**: Share with peers or instructor

### 💡 Going Further:
- Try this workflow on a different dataset
- Experiment with deep learning models
- Deploy your model as a web application
- Participate in Kaggle competitions

---

**Great work! You're now ready for more advanced ML projects! 🎓**